[WIP] Variance tracking for DTensor manual APIs

vishal9-team opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #173114
* #173107
* __->__ #169620

**Design Document: [gdoc](https://docs.google.com/document/d/1vi0UD-uaow3lZXWnnBlZIw_4011TS44HsSBS-bUkQwU/edit?tab=t.0#heading=h.4cu9dagbd0bu)**

Use cases below previously required precise Placements to be specified by user via ```in_grad_placements``` parameter in local_map

### Use Case 1: Data Parallel:  [**_test_data_parallel_**]
Inserts all_reduce in backward for Parameters

<img width="480" height="480" alt="data_parallel" src="https://github.com/user-attachments/assets/075abb21-8f52-421d-b8c1-21df0fefae92" />




### Use Case 2: Megatron Tensor Parallel: [**_test_megatron_tp_parallel_**]

Algorithm: https://arxiv.org/pdf/1909.08053

<img width="480" height="480" alt="megatron_tp" src="https://github.com/user-attachments/assets/c79eb229-37e1-4d2e-8174-0ab5df97b7d5" />

Inserts all_reduce in backward for Input and Parameters as appropriate

<img width="480" height="720" alt="graph_ltensor_megatron_tp" src="https://github.com/user-attachments/assets/ba1cfe56-c431-44da-915c-da60b8cfa554" />

### Use Case 3: FSDP + TP: [**_test_megatron_tp_parallel_**]

TP requires all_reduce in backward for input and FSDP requires reduce_scatter for parameters - both are a correctly added

<img width="1152" height="678" alt="graph_ltensor_rs" src="https://github.com/user-attachments/assets/dc95901a-2745-40c9-8b5c-4751529b48dd" />


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo

vishal9-team committed:
[WIP] Variance tracking for DTensor manual APIs

[ghstack-poisoned]

vishal9-team committed:
Update on "[WIP: Early Draft] Variance tracking for DTensor manual APIs"




[ghstack-poisoned]

dzmitry-huba commented on test/distributed/test_ltensor_variance_tracking.py:
```diff
@@ -0,0 +1,182 @@
+import torch
+import torch.distributed as dist
+from torch.distributed.device_mesh import DeviceMesh
+from torch.distributed.tensor.ltensor.ltensor import LTensor
```
If anything, I'm to blame for the LTensor name :) Vishal asked me if LocalTensor can be renamed and I suggested to use LTensor name (L for Local similar D for distributed). If people are feeling strongly about renaming LocalTensor, let's discuss it before taking renaming (there are quite a few places)

vishal9-team committed:
Update on "[WIP: Early Draft] Variance tracking for DTensor manual APIs"




[ghstack-poisoned]

vishal9-team committed:
Update on "[WIP: Early Draft] Variance tracking for DTensor manual APIs"




[ghstack-poisoned]

albanD commented on torch/distributed/tensor/_ltensor.py:
```diff
@@ -0,0 +1,187 @@
+            if not missing_axes:
+                return local_tensor
+
+            from . import _varying_collectives as vcols
```
import at the top

vishal9-team committed:
Update on "[WIP: Early Draft] Variance tracking for DTensor manual APIs"


Use cases below previously required precise Placements to be specified by user via ```in_grad_placements``` parameter in local_map

### Use Case 1: Data Parallel:  [**_test_data_parallel_**]
Inserts all_reduce in backward for Parameters

<img width="480" height="480" alt="data_parallel" src="https://github.com/user-attachments/assets/075abb21-8f52-421d-b8c1-21df0fefae92" />



### Use Case 2: Simple Tensor Parallel: [**_test_tensor_parallel_**]

Inserts all_reduce in backward for Input

<img width="480" height="480" alt="tensor_parallel" src="https://github.com/user-attachments/assets/b8ae5f51-99ff-4e04-827b-5fed328f3eac" />


[ghstack-poisoned]

vishal9-team committed:
Update on "[WIP: Early Draft] Variance tracking for DTensor manual APIs"


**Design Document: [gdoc](https://docs.google.com/document/d/1vi0UD-uaow3lZXWnnBlZIw_4011TS44HsSBS-bUkQwU/edit?tab=t.0#heading=h.4cu9dagbd0bu)**

Use cases below previously required precise Placements to be specified by user via ```in_grad_placements``` parameter in local_map

### Use Case 1: Data Parallel:  [**_test_data_parallel_**]
Inserts all_reduce in backward for Parameters

<img width="480" height="480" alt="data_parallel" src="https://github.com/user-attachments/assets/075abb21-8f52-421d-b8c1-21df0fefae92" />




### Use Case 2: Megatron Tensor Parallel: [**_test_megatron_tp_parallel_**]

Algorithm: https://arxiv.org/pdf/1909.08053
<img width="480" height="480" alt="megatron_tp" src="https://github.com/user-attachments/assets/c79eb229-37e1-4d2e-8174-0ab5df97b7d5" />

Inserts all_reduce in backward for Input and Parameters as appropriate

<img width="480" height="720" alt="graph_ltensor_megatron_tp" src="https://github.com/user-attachments/assets/ba1cfe56-c431-44da-915c-da60b8cfa554" />


[ghstack-poisoned]

vishal9-team committed:
Update on "[WIP] Variance tracking for DTensor manual APIs"


**Design Document: [gdoc](https://docs.google.com/document/d/1vi0UD-uaow3lZXWnnBlZIw_4011TS44HsSBS-bUkQwU/edit?tab=t.0#heading=h.4cu9dagbd0bu)**

Use cases below previously required precise Placements to be specified by user via ```in_grad_placements``` parameter in local_map

### Use Case 1: Data Parallel:  [**_test_data_parallel_**]
Inserts all_reduce in backward for Parameters

<img width="480" height="480" alt="data_parallel" src="https://github.com/user-attachments/assets/075abb21-8f52-421d-b8c1-21df0fefae92" />




### Use Case 2: Megatron Tensor Parallel: [**_test_megatron_tp_parallel_**]

Algorithm: https://arxiv.org/pdf/1909.08053
<img width="480" height="480" alt="megatron_tp" src="https://github.com/user-attachments/assets/c79eb229-37e1-4d2e-8174-0ab5df97b7d5" />

Inserts all_reduce in backward for Input and Parameters as appropriate

<img width="480" height="720" alt="graph_ltensor_megatron_tp" src="https://github.com/user-attachments/assets/ba1cfe56-c431-44da-915c-da60b8cfa554" />


[ghstack-poisoned]

vishal9-team commented:
@pytorchmergebot rebase

vishal9-team committed:
Update on "[WIP] Variance tracking for DTensor manual APIs"


**Design Document: [gdoc](https://docs.google.com/document/d/1vi0UD-uaow3lZXWnnBlZIw_4011TS44HsSBS-bUkQwU/edit?tab=t.0#heading=h.4cu9dagbd0bu)**

Use cases below previously required precise Placements to be specified by user via ```in_grad_placements``` parameter in local_map

### Use Case 1: Data Parallel:  [**_test_data_parallel_**]
Inserts all_reduce in backward for Parameters

<img width="480" height="480" alt="data_parallel" src="https://github.com/user-attachments/assets/075abb21-8f52-421d-b8c1-21df0fefae92" />




### Use Case 2: Megatron Tensor Parallel: [**_test_megatron_tp_parallel_**]

Algorithm: https://arxiv.org/pdf/1909.08053

<img width="480" height="480" alt="megatron_tp" src="https://github.com/user-attachments/assets/c79eb229-37e1-4d2e-8174-0ab5df97b7d5" />

Inserts all_reduce in backward for Input and Parameters as appropriate

<img width="480" height="720" alt="graph_ltensor_megatron_tp" src="https://github.com/user-attachments/assets/ba1cfe56-c431-44da-915c-da60b8cfa554" />

### Use Case 3: FSDP + TP: [**_test_megatron_tp_parallel_**]

TP requires all_reduce in backward for input and FSDP requires reduce_scatter for parameters - both are a correctly added

<img width="1152" height="678" alt="graph_ltensor_rs" src="https://github.com/user-attachments/assets/dc95901a-2745-40c9-8b5c-4751529b48dd" />


[ghstack-poisoned]

vishal9-team committed:
Update on "[WIP] Variance tracking for DTensor manual APIs"


**Design Document: [gdoc](https://docs.google.com/document/d/1vi0UD-uaow3lZXWnnBlZIw_4011TS44HsSBS-bUkQwU/edit?tab=t.0#heading=h.4cu9dagbd0bu)**

Use cases below previously required precise Placements to be specified by user via ```in_grad_placements``` parameter in local_map

### Use Case 1: Data Parallel:  [**_test_data_parallel_**]
Inserts all_reduce in backward for Parameters

<img width="480" height="480" alt="data_parallel" src="https://github.com/user-attachments/assets/075abb21-8f52-421d-b8c1-21df0fefae92" />




### Use Case 2: Megatron Tensor Parallel: [**_test_megatron_tp_parallel_**]

Algorithm: https://arxiv.org/pdf/1909.08053

<img width="480" height="480" alt="megatron_tp" src="https://github.com/user-attachments/assets/c79eb229-37e1-4d2e-8174-0ab5df97b7d5" />

Inserts all_reduce in backward for Input and Parameters as appropriate

<img width="480" height="720" alt="graph_ltensor_megatron_tp" src="https://github.com/user-attachments/assets/ba1cfe56-c431-44da-915c-da60b8cfa554" />

### Use Case 3: FSDP + TP: [**_test_megatron_tp_parallel_**]

TP requires all_reduce in backward for input and FSDP requires reduce_scatter for parameters - both are a correctly added

<img width="1152" height="678" alt="graph_ltensor_rs" src="https://github.com/user-attachments/assets/dc95901a-2745-40c9-8b5c-4751529b48dd" />


[ghstack-poisoned]

vishal9-team committed:
Update on "[WIP] Variance tracking for DTensor manual APIs"


**Design Document: [gdoc](https://docs.google.com/document/d/1vi0UD-uaow3lZXWnnBlZIw_4011TS44HsSBS-bUkQwU/edit?tab=t.0#heading=h.4cu9dagbd0bu)**

Use cases below previously required precise Placements to be specified by user via ```in_grad_placements``` parameter in local_map

### Use Case 1: Data Parallel:  [**_test_data_parallel_**]
Inserts all_reduce in backward for Parameters

<img width="480" height="480" alt="data_parallel" src="https://github.com/user-attachments/assets/075abb21-8f52-421d-b8c1-21df0fefae92" />




### Use Case 2: Megatron Tensor Parallel: [**_test_megatron_tp_parallel_**]

Algorithm: https://arxiv.org/pdf/1909.08053

<img width="480" height="480" alt="megatron_tp" src="https://github.com/user-attachments/assets/c79eb229-37e1-4d2e-8174-0ab5df97b7d5" />

Inserts all_reduce in backward for Input and Parameters as appropriate

<img width="480" height="720" alt="graph_ltensor_megatron_tp" src="https://github.com/user-attachments/assets/ba1cfe56-c431-44da-915c-da60b8cfa554" />

### Use Case 3: FSDP + TP: [**_test_megatron_tp_parallel_**]

TP requires all_reduce in backward for input and FSDP requires reduce_scatter for parameters - both are a correctly added

<img width="1152" height="678" alt="graph_ltensor_rs" src="https://github.com/user-attachments/assets/dc95901a-2745-40c9-8b5c-4751529b48dd" />


[ghstack-poisoned]

vishal9-team committed:
Update on "[WIP] Variance tracking for DTensor manual APIs"


**Design Document: [gdoc](https://docs.google.com/document/d/1vi0UD-uaow3lZXWnnBlZIw_4011TS44HsSBS-bUkQwU/edit?tab=t.0#heading=h.4cu9dagbd0bu)**

Use cases below previously required precise Placements to be specified by user via ```in_grad_placements``` parameter in local_map

### Use Case 1: Data Parallel:  [**_test_data_parallel_**]
Inserts all_reduce in backward for Parameters

<img width="480" height="480" alt="data_parallel" src="https://github.com/user-attachments/assets/075abb21-8f52-421d-b8c1-21df0fefae92" />




### Use Case 2: Megatron Tensor Parallel: [**_test_megatron_tp_parallel_**]

Algorithm: https://arxiv.org/pdf/1909.08053

<img width="480" height="480" alt="megatron_tp" src="https://github.com/user-attachments/assets/c79eb229-37e1-4d2e-8174-0ab5df97b7d5" />

Inserts all_reduce in backward for Input and Parameters as appropriate

<img width="480" height="720" alt="graph_ltensor_megatron_tp" src="https://github.com/user-attachments/assets/ba1cfe56-c431-44da-915c-da60b8cfa554" />

### Use Case 3: FSDP + TP: [**_test_megatron_tp_parallel_**]

TP requires all_reduce in backward for input and FSDP requires reduce_scatter for parameters - both are a correctly added

<img width="1152" height="678" alt="graph_ltensor_rs" src="https://github.com/user-attachments/assets/dc95901a-2745-40c9-8b5c-4751529b48dd" />


[ghstack-poisoned]

vishal9-team committed:
Update on "[WIP] Variance tracking for DTensor manual APIs"


**Design Document: [gdoc](https://docs.google.com/document/d/1vi0UD-uaow3lZXWnnBlZIw_4011TS44HsSBS-bUkQwU/edit?tab=t.0#heading=h.4cu9dagbd0bu)**

Use cases below previously required precise Placements to be specified by user via ```in_grad_placements``` parameter in local_map

### Use Case 1: Data Parallel:  [**_test_data_parallel_**]
Inserts all_reduce in backward for Parameters

<img width="480" height="480" alt="data_parallel" src="https://github.com/user-attachments/assets/075abb21-8f52-421d-b8c1-21df0fefae92" />




### Use Case 2: Megatron Tensor Parallel: [**_test_megatron_tp_parallel_**]

Algorithm: https://arxiv.org/pdf/1909.08053

<img width="480" height="480" alt="megatron_tp" src="https://github.com/user-attachments/assets/c79eb229-37e1-4d2e-8174-0ab5df97b7d5" />

Inserts all_reduce in backward for Input and Parameters as appropriate

<img width="480" height="720" alt="graph_ltensor_megatron_tp" src="https://github.com/user-attachments/assets/ba1cfe56-c431-44da-915c-da60b8cfa554" />

### Use Case 3: FSDP + TP: [**_test_megatron_tp_parallel_**]

TP requires all_reduce in backward for input and FSDP requires reduce_scatter for parameters - both are a correctly added

<img width="1152" height="678" alt="graph_ltensor_rs" src="https://github.com/user-attachments/assets/dc95901a-2745-40c9-8b5c-4751529b48dd" />


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

vishal9-team committed:
Update on "[WIP] Variance tracking for DTensor manual APIs"


**Design Document: [gdoc](https://docs.google.com/document/d/1vi0UD-uaow3lZXWnnBlZIw_4011TS44HsSBS-bUkQwU/edit?tab=t.0#heading=h.4cu9dagbd0bu)**

Use cases below previously required precise Placements to be specified by user via ```in_grad_placements``` parameter in local_map

### Use Case 1: Data Parallel:  [**_test_data_parallel_**]
Inserts all_reduce in backward for Parameters

<img width="480" height="480" alt="data_parallel" src="https://github.com/user-attachments/assets/075abb21-8f52-421d-b8c1-21df0fefae92" />




### Use Case 2: Megatron Tensor Parallel: [**_test_megatron_tp_parallel_**]

Algorithm: https://arxiv.org/pdf/1909.08053

<img width="480" height="480" alt="megatron_tp" src="https://github.com/user-attachments/assets/c79eb229-37e1-4d2e-8174-0ab5df97b7d5" />

Inserts all_reduce in backward for Input and Parameters as appropriate

<img width="480" height="720" alt="graph_ltensor_megatron_tp" src="https://github.com/user-attachments/assets/ba1cfe56-c431-44da-915c-da60b8cfa554" />

### Use Case 3: FSDP + TP: [**_test_megatron_tp_parallel_**]

TP requires all_reduce in backward for input and FSDP requires reduce_scatter for parameters - both are a correctly added

<img width="1152" height="678" alt="graph_ltensor_rs" src="https://github.com/user-attachments/assets/dc95901a-2745-40c9-8b5c-4751529b48dd" />


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

vishal9-team committed:
Update on "[WIP] Variance tracking for DTensor manual APIs"


**Design Document: [gdoc](https://docs.google.com/document/d/1vi0UD-uaow3lZXWnnBlZIw_4011TS44HsSBS-bUkQwU/edit?tab=t.0#heading=h.4cu9dagbd0bu)**

Use cases below previously required precise Placements to be specified by user via ```in_grad_placements``` parameter in local_map

### Use Case 1: Data Parallel:  [**_test_data_parallel_**]
Inserts all_reduce in backward for Parameters

<img width="480" height="480" alt="data_parallel" src="https://github.com/user-attachments/assets/075abb21-8f52-421d-b8c1-21df0fefae92" />




### Use Case 2: Megatron Tensor Parallel: [**_test_megatron_tp_parallel_**]

Algorithm: https://arxiv.org/pdf/1909.08053

<img width="480" height="480" alt="megatron_tp" src="https://github.com/user-attachments/assets/c79eb229-37e1-4d2e-8174-0ab5df97b7d5" />

Inserts all_reduce in backward for Input and Parameters as appropriate

<img width="480" height="720" alt="graph_ltensor_megatron_tp" src="https://github.com/user-attachments/assets/ba1cfe56-c431-44da-915c-da60b8cfa554" />

### Use Case 3: FSDP + TP: [**_test_megatron_tp_parallel_**]

TP requires all_reduce in backward for input and FSDP requires reduce_scatter for parameters - both are a correctly added

<img width="1152" height="678" alt="graph_ltensor_rs" src="https://github.com/user-attachments/assets/dc95901a-2745-40c9-8b5c-4751529b48dd" />


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

vishal9-team committed:
Update on "[WIP] Variance tracking for DTensor manual APIs"


**Design Document: [gdoc](https://docs.google.com/document/d/1vi0UD-uaow3lZXWnnBlZIw_4011TS44HsSBS-bUkQwU/edit?tab=t.0#heading=h.4cu9dagbd0bu)**

Use cases below previously required precise Placements to be specified by user via ```in_grad_placements``` parameter in local_map

### Use Case 1: Data Parallel:  [**_test_data_parallel_**]
Inserts all_reduce in backward for Parameters

<img width="480" height="480" alt="data_parallel" src="https://github.com/user-attachments/assets/075abb21-8f52-421d-b8c1-21df0fefae92" />




### Use Case 2: Megatron Tensor Parallel: [**_test_megatron_tp_parallel_**]

Algorithm: https://arxiv.org/pdf/1909.08053

<img width="480" height="480" alt="megatron_tp" src="https://github.com/user-attachments/assets/c79eb229-37e1-4d2e-8174-0ab5df97b7d5" />

Inserts all_reduce in backward for Input and Parameters as appropriate

<img width="480" height="720" alt="graph_ltensor_megatron_tp" src="https://github.com/user-attachments/assets/ba1cfe56-c431-44da-915c-da60b8cfa554" />

### Use Case 3: FSDP + TP: [**_test_megatron_tp_parallel_**]

TP requires all_reduce in backward for input and FSDP requires reduce_scatter for parameters - both are a correctly added

<img width="1152" height="678" alt="graph_ltensor_rs" src="https://github.com/user-attachments/assets/dc95901a-2745-40c9-8b5c-4751529b48dd" />


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

vishal9-team committed:
Update on "[WIP] Variance tracking for DTensor manual APIs"


**Design Document: [gdoc](https://docs.google.com/document/d/1vi0UD-uaow3lZXWnnBlZIw_4011TS44HsSBS-bUkQwU/edit?tab=t.0#heading=h.4cu9dagbd0bu)**

Use cases below previously required precise Placements to be specified by user via ```in_grad_placements``` parameter in local_map

### Use Case 1: Data Parallel:  [**_test_data_parallel_**]
Inserts all_reduce in backward for Parameters

<img width="480" height="480" alt="data_parallel" src="https://github.com/user-attachments/assets/075abb21-8f52-421d-b8c1-21df0fefae92" />




### Use Case 2: Megatron Tensor Parallel: [**_test_megatron_tp_parallel_**]

Algorithm: https://arxiv.org/pdf/1909.08053

<img width="480" height="480" alt="megatron_tp" src="https://github.com/user-attachments/assets/c79eb229-37e1-4d2e-8174-0ab5df97b7d5" />

Inserts all_reduce in backward for Input and Parameters as appropriate

<img width="480" height="720" alt="graph_ltensor_megatron_tp" src="https://github.com/user-attachments/assets/ba1cfe56-c431-44da-915c-da60b8cfa554" />

### Use Case 3: FSDP + TP: [**_test_megatron_tp_parallel_**]

TP requires all_reduce in backward for input and FSDP requires reduce_scatter for parameters - both are a correctly added

<img width="1152" height="678" alt="graph_ltensor_rs" src="https://github.com/user-attachments/assets/dc95901a-2745-40c9-8b5c-4751529b48dd" />


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

vishal9-team committed:
Update on "[WIP] Variance tracking for DTensor manual APIs"


**Design Document: [gdoc](https://docs.google.com/document/d/1vi0UD-uaow3lZXWnnBlZIw_4011TS44HsSBS-bUkQwU/edit?tab=t.0#heading=h.4cu9dagbd0bu)**

Use cases below previously required precise Placements to be specified by user via ```in_grad_placements``` parameter in local_map

### Use Case 1: Data Parallel:  [**_test_data_parallel_**]
Inserts all_reduce in backward for Parameters

<img width="480" height="480" alt="data_parallel" src="https://github.com/user-attachments/assets/075abb21-8f52-421d-b8c1-21df0fefae92" />




### Use Case 2: Megatron Tensor Parallel: [**_test_megatron_tp_parallel_**]

Algorithm: https://arxiv.org/pdf/1909.08053

<img width="480" height="480" alt="megatron_tp" src="https://github.com/user-attachments/assets/c79eb229-37e1-4d2e-8174-0ab5df97b7d5" />

Inserts all_reduce in backward for Input and Parameters as appropriate

<img width="480" height="720" alt="graph_ltensor_megatron_tp" src="https://github.com/user-attachments/assets/ba1cfe56-c431-44da-915c-da60b8cfa554" />

### Use Case 3: FSDP + TP: [**_test_megatron_tp_parallel_**]

TP requires all_reduce in backward for input and FSDP requires reduce_scatter for parameters - both are a correctly added

<img width="1152" height="678" alt="graph_ltensor_rs" src="https://github.com/user-attachments/assets/dc95901a-2745-40c9-8b5c-4751529b48dd" />


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]
